de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/negotiation/ProjectNegotiation.java

```diff
@@ -155,4 +161,17 @@ protected void monitorFileTransfer(FileTransfer transfer, IProgressMonitor monit
   protected void notifyTerminated(NegotiationListener listener) {
     listener.negotiationTerminated(this);
   }
+
+  /**
+   * Returns the JID of the peer with which the negotiation takes place.
+   *
+   * @return peer JID
+   */
+  public JID getPeer() {
```

Comment:
Now ProjectNegotiation can store User final, moving `setPeer` to it's only caller and no additional JID is stored.

Reply:
This is still duplicated code.

As the getPeer method was not marked as final you could use override it in the affected class, i.e make the JID private final in the Negotiation class and let getPeer return this final JID. This would at least 3 classes work as they should. You could then hack together you own stuff in the affected class.

As for the JID stuff. If you can ensure that the correct fully qualified JID is passed in from the UI (the one with Saros) you could drop the JID query. I have my doubt that Saros/E will present you the correct JID all the time and I have no clue what Saros/I is passing in.

Reply:
?

private final JID jid;
public Negotiation(final JID jid ,,,){this.jid = jid} public getPeer() {return jid}

private final User user;
public ProjectNegotiation(final User user ...) {super(user.getJID()); this.user = user} public getRemoteUser() {return user}

Reply:
Do you even need getPeer() in the ProjectNegotiation ?

Reply:
> private final JID jid;

You can't make this final without changing `OutgoingSessionNegotiation` and I don't understand where you see the advantage of keeping jid in Negotiation.

> Do you even need getPeer() in the ProjectNegotiation ?

not in ProjectNegotiation but didn't want to change dependent classes now. Can mark it deprecated if it helps.